// koishi-plugin-cron/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apply = exports.Config = exports.name = void 0;
const koishi_1 = require("koishi");
const cron_parser_1 = require("cron-parser");
class Task {
    constructor(ctx, expr, callback) {
        this.ctx = ctx;
        this.expr = expr;
        this.callback = callback;
        this.start();
    }
    start() {
        this.timer = setTimeout(async () => {
            this.start();
            try {
                await this.callback();
            }
            catch (error) {
                this.ctx.logger('cron').warn(error);
            }
        }, this.expr.next().getTime() - Date.now());
    }
    stop() {
        clearTimeout(this.timer);
    }
}
exports.name = 'cron';
exports.Config = koishi_1.Schema.object({});
function apply(ctx) {
    ctx.provide('cron');
    ctx.cron = function cron(input, callback) {
        const task = new Task(this, (0, cron_parser_1.parseExpression)(input), callback);
        return this.collect('cron', () => task.stop());
    };
}
exports.apply = apply;
